### PR TITLE
Add SpatialExperiment methods to rotate/mirror spatialCoords, as well…

### PR DESCRIPTION
**Pull request by Nick-Eagles**
… as wrappers to rotate/mirror the entire SpatialExperiment (spatialCoords and imgData)

# List of individual changes that were performed

* added `rotateCoords` function to `SpatialExperiment-methods.R` , not exported
* added `mirrorCoords` function to `SpatialExperiment-methods.R` , not exported
* added `rotateObject` function to `SpatialExperiment-methods.R`
* added `mirrorObject` function to `SpatialExperiment-methods.R`
* `rotateCoords` unit test: non-null `sample_id`
* `mirrorCoords` unit test: identity mirror
* `mirrorCoords` unit test: non-null `sample_id`
* `rotateObject` unit test: rotations by angles adding to 0 have no effect
* `mirrorObject` unit test: identity transformations
* `mirrorObject` unit test: symmetric composition of constituent functions
* added `@aliases` for all functions
* modified `@description` to include new function overviews
* added `@section` spatialCoords transformation methods
* added `@section` SpatialExperiment transformation wrapper methods
* added `@params` for `degrees` and `axis`
* added `@examples` for all functions

# Rationale behind some choices

Note that I chose not to export `rotateCoords` and `mirrorCoords`, since problems can result from rotating just the `spatialCoords` or just the `imgData`. For the same reason, if this pull request is accepted, I believe we should not allow users to call `rotateImg` directly (rather, only call the wrapper function `rotateObject` that I wrote). To describe the potential "problems" in more detail:

If you're interested in rotating `spatialCoords` and then rotating the `SpatialImage` by the same angle, while maintaining where the coordinates lie relative to the image (functions like `ggspavis::plotVisium` display both superimposed), the shape of the `imgRaster` must be used.  This is because [.rotate](https://github.com/drighelli/SpatialExperiment/blob/25c25cd9747279c5dd293b2aa5e12398e92bf851/R/SpatialImage-utils.R#L52) relies on transposing and reversing elements of the `imgRaster`, operations whose results depend on the image dimensions. This is problematic because if users are allowed to freely call `rotateImg` or `rotateCoords` independently, this creates a mismatch between the `spatialCoords` and `imgRaster`. Then subsequent calls to `rotateCoords` cannot possibly perform the rotation consistently with `rotateImg` (unless the starting shape of the `SpatialImage` before any rotations is somehow saved). Note that this problem occurs regardless of the rotation method used (e.g. matrix multiplication or via `.rotate`).
Basically, given how `rotateImg` is implemented (via `.rotate`), a potential `rotateCoords` method doesn't have enough information to operate as expected in general. Restricting users to using only `rotateObject` instead of `rotateImg` or `rotateCoords` prevents any mismatch and resulting problems.

My changes didn't modify existing functions, and therefore I left `rotateImg` exported as it was, despite my recommendations above.

Sorry for the wall of text! I'm happy to hear thoughts or make additional edits.

Best,
-Nick

### CODE REVIEW

**Comment by lcolladotor** [comment]
Great work Nick! Thanks for your first PR ^^

**Comment by lcolladotor** [comment]
If you prefer that we schedule a short Zoom session to talk about the PR, please let us know and Nick can help coordinate the when2meet.

Anyways, thanks again for checking this PR!

Best,
Leo

**Comment by lcolladotor** [comment]
@Nick-Eagles is this PR up to date? Sounds like we'll have to use these functions soon on both the HPC and the NAc projects we are working on / will be working on.

**Comment by lmweber** [comment]
Thanks for the reminder -- apologies we haven't reviewed / merged this yet

**Comment by Nick-Eagles** [comment]
Hello, just following up on this PR-- looks like I'll have to resolve some conflicts before a merge is possible, but is there still interest in this feature? Thanks

**Comment by drighelli** [comment]
Hi Nick @Nick-Eagles ,

sorry for my very late answer, I looked into your code and I think that's a very good job.

On the other hand, I understand why you don't want to export `rotateImg` and `rotateCoords` independently, but I would say that it's up to the user not to do that.

I don't know if this could be ever possible but, let's imagine we have a mismatch from the machine, which could export correct positions in terms of coordinates, but a rotated image (or vice-versa), the final user would like to match the two things.
So I'd suggest exporting these functions, but with a warning message with each of them, something like "Please be aware to rotate the image/coords accordingly, or use the rotateObject method instead".

About your last question: yes I think this could be really useful, so proceed with solving the conflicts, and please add yourself as a contributor in the DESCRIPTION file.

Also, I'd place all the rotating/mirroring methods in a separate file.

Thank you very much and sorry again for taking so much time for reviewing this PR.

Dario

File: R/SpatialExperiment-methods.R [hunk 1]
```diff
@@ -192,6 +249,173 @@ setMethod("spatialCoords",
     "SpatialExperiment",
     function(x) int_colData(x)$spatialCoords)
 
+#' @rdname SpatialExperiment-methods
```
Comment (drighelli) [inline R/SpatialExperiment-methods.R]: maybe we want to export this method and add a warning message with it.

Also, it could be a good idea to place rotation methods in a separate file

**Comment by Nick-Eagles** [comment]
Thanks for the feedback. I think the PR should be good to go now. I:

- resolved merge conflicts (which was almost a clean fast-forward, only having a single line of code I needed to incorporate)
- exported `rotateCoords` and `mirrorCoords` but added a parameter `warn` that defaults to `TRUE`, which prints a descriptive warning about possible issues. It can be set to `FALSE` to suppress the warnings, and the wrappers `rotateObject` and `mirrorObject` invoke them with `warn = FALSE`.
- added myself as a contributor
- fixed a missing variable (`dim_max`) definition in `mirrorCoords`. I also had to explicitly coerce the columns of `spatialCoords` to be integers (as was done in `rotateCoords`) rather than rounded numeric vectors. Not sure how these things slipped through my initial testing + `testthat` unit tests, but all unit tests now pass.

I haven't placed my new methods in a separate file, as I wasn't sure about naming and organization. Would you prefer `rotateCoords`, `mirrorCoords`, `rotateObject`, and `mirrorObject` in a separate file? In my mind, `rotateImg` and `mirrorImg` are part of the same conceptual grouping of functionality as well.

**Comment by lmweber** [comment]
Hi @Nick-Eagles , thanks for this pull request, and apologies also for the slow response.

This looks great, and I am running some local checks now. I agree that exporting `rotate/mirrorCoords()` with the warning message makes sense, since this gives users flexibility if they have a special case where they want to run these individually.

Thanks also for including the unit tests.

I also think it would make sense to put the new methods in a separate file, e.g. named `SpatialExperiment-rotate-mirror.R`, for easier long-term maintenance (since the existing file `SpatialExperiment-methods.R` is already quite long), although as you pointed out this creates some inconsistency with `rotate/mirrorImg()`. I'll look at this some more now -- maybe it is ok either way.

**Comment by lmweber** [comment]
I'm getting some errors in the checks regarding the generics and in the documentation -- I'll do some checking and will add some commits here in the PR if I figure it out and find anything missing.

**Comment by lmweber** [comment]
@Nick-Eagles @drighelli I have added the following updates below in the commits above.

Could you please check these (especially the one about the generics)? I will also do some more checking and add some further minor updates, e.g. NEWS item and version bump. Thank you!

- added generics for the new rotate / mirror functions (this was returning an error since these were missing)
- removed a line in the documentation code block with redundant inherited parameters (this was returning a warning)
- compiled documentation

**Comment by lmweber** [comment]
Hi, I added some more commits to do the following:

- split the .R files so we have a separate .R file for the new rotate / mirror functions, since the existing one (`SpatialExperiment-methods.R`) was getting very long
- similarly for the unit tests, split the files so we have a separate file with the unit tests for the rotate / mirror functions
- add NEWS
- add version bump

@drighelli could you please have a look (especially to check that the way I set up the generic functions was correct). @Nick-Eagles if you have any comments on my changes / commits above please let me know too. Otherwise I think we are ok to merge and push to Bioc-devel. Thanks again for the PR @Nick-Eagles .

**Comment by Nick-Eagles** [comment]
Thanks! Keeping these things in a separate file is totally fine, and the other minor adjustments you made seem reasonable to me.

**Comment by lmweber** [comment]
Great, thanks for checking @Nick-Eagles !

**Review by drighelli — APPROVED** [review]

**Comment by drighelli** [comment]
Thanks guys (@Nick-Eagles, @lmweber), everything looks good to me!

I proceeded with the merging, Lukas, if you have time, could you push the changes to the bioc devel branch?

Otherwise, I'll take care of it. :)

**Comment by lmweber** [comment]
Great, thank you! Yes sounds good, I'll sync everything and push to Bioc-devel (upstream) now.
